File: code/sorting/src/binary_insertion_sort/binary_insertion_sort.py
# Binary insertion sort finds the appropriate index to place element using binary search
# it has less complexity as compared to insertion sort
def BinarySearch(l, k, p, q):
    # l is the arguement used that contains the unsorted element
    # k variable contains the elements which is to be inserted
    # p variable contains the index of first element from unsorted array
    # q contains the index of the last element from unsorted array
    if p == q:
        if l[p] > k:
            return p
        else:
            return p + 1
    if p > q:
        return p
    # m contains the index of middle element from array
    m = int((p + q) / 2)
    # To check if middle element is smaller than selected element
    if l[m] < k:
        return BinarySearch(l, k, m + 1, q)
    # To check if middle element is greater than selected element
    elif l[m] > k:
        return BinarySearch(l, k, p, m - 1)
    # if middle element is equal to selected element
    else:
        return m


# InsertionSort function traverses via the unsorted array
def InsertionSort(l):
    # for loop contains the array starting from index 1
    for i in range(1, len(l)):
        k = l[i]
        # BinarySearch function finds the appropriate index for the element
        j = BinarySearch(l, k, 0, i - 1)
        # once the index is found, Element is placed at the index in an array
        l = l[:j] + [k] + l[j:i] + l[i + 1 :]
    return l


# To display the sorted array
print("New Array:")
l = [37, 23, 0, 17, 12, 72, 31, 46, 100, 88, 54]
print(InsertionSort(l))

# output->0 12 17 23 31 37 46 54 72 88 100

File: code/sorting/src/binary_insertion_sort/binary_insertion_sort.c
#include <stdio.h> 
/* Binary Insertion sort is used to find index to place element in array. */
/* It has less complexity as compared to Insertion sort */
int BinarySearch(int l[], int value, int start, int end) 
{ 
	/* l is the array that is to be sorted */
	/* Value variable contains elements from array that is to be inserted */
	/* start variable contains the index of first element from unsorted array */
	/* end variable stores index of last element from unsorted array */
	if (end <= start) 
		return ((value > l[start]) ? (start + 1) : start); 
	/* m contains the middle element index of an array */
	int m = (start + end)/2; 
	/* To check if selected element is equal to middle element in array */
	if(value == l[m])
		return (m + 1); 
	/* To check if selected element is greater than the middle element in array */
	if(value > l[m])
		return (BinarySearch(l, value, m + 1, end));
	return (BinarySearch(l, value, start, m - 1));
} 

/* InsetionSort Funciton is used to sort an array l of size n */
void InsertionSort(int l[], int n) 
{ 	
	/* l argument contains unsorted array */
	/* n is the length of array */
	int i, loc, j, k, value; 
	/* for loop traverses the unsorted array */
	for (i = 1; i < n; ++i) 
	{ 
		j = i - 1; 
		value = l[i]; 

		/* BinarySearch Function is called to find the index in an array */
		loc = BinarySearch(l, value, 0, j); 

		/* Elements are shifted to their indexes */
		while (j >= loc) 
		{ 
			l[j + 1] = l[j]; 
			j--; 
		} 
		l[j + 1] = value; 
	} 
} 

/* Main Program to create a list of elements to be sorted */
int main() 
{ 
	int l[] = {37, 23, 0, 17, 12, 72, 31, 46, 100, 88, 54}; 
	/* n is the length of the unsorted array */
	int n = sizeof(l)/sizeof(l[0]); 

	InsertionSort(l, n); 
	/* for loop displays the sorted array elements */
	printf("Sorted array: "); 
	for (int i = 0; i < n; ++i) 
		printf("%d ",l[i]); 

	return (0); 
} 
/* Sorted array: 0 12 17 23 31 37 46 54 72 88 100 */
